google/cloud/security/scanner/audit/fw_rules_engine.py

```diff
@@ -14,64 +14,354 @@
+        if rule_defs:
+            self.add_rules(rule_defs)
+        if group_defs:
+            self.add_rule_groups(group_defs)
+        if org_policy:
+            self.add_org_policy(org_policy)
+
+    def add_rules(self, rule_defs):
+        """Adds rules to rule book.
+
+        Args:
+          rule_defs (list): Rule definition dictionaries from yaml config file.
+
+        Raises:
+          InvalidRuleDefinition: If the rule is missing required fields or the
+            fields have invalid values.
+        """
+        self._rules_sema.acquire() #TODO change to lock
+        for i, rule_def in enumerate(rule_defs):
+            if rule_def is not None:
+                self.add_rule(rule_def, i)
+        self._rules_sema.release()
+
+    def add_rule(self, rule_def, rule_index):
+        """Adds a rule to the rule book.
+
+        Args:
+          rule_def (Rule): A Rule used to check for violations.
+          rule_index (int): Used for logs.
+
+        Raises:
+          DuplicateFirewallRuleError: When the rule by the same name exists.
+        """
+        rule = Rule.from_config(rule_def)
+        if rule.id in self.rules_map:
+            raise DuplicateFirewallRuleError(
+                'Rule id "%s" already in rules (rule %s)' % (
+                    rule.id, rule_index))
+        self.rules_map[rule.id] = rule
+
+    def add_rule_groups(self, group_defs):
+        """Creates group to rule matching.
+
+        Args:
+          group_defs (dict): A dictionary with a group id and a list of rule ids
+            that will be included by including this group in a policy.
+
+        Raises:
+          DuplicateFirewallGroupError: Raised if the group id already exists.
+          RuleDoesntExistError: Raised if a rule included in the group does not
+            exist.
+          InvalidGroupDefinition: Raised if a group definition is invalid.
+        """
+        for group_def in group_defs:
+            group_id = group_def.get('group_id')
+            if not group_id:
+                raise InvalidGroupDefinition('Group requires a group id')
+            if group_id in self.rule_groups_map:
+                raise DuplicateFirewallGroupError(
+                    'Group id already exists: %s' % group_id)
+            rule_ids = group_def.get('rule_ids')
+            if not rule_ids:
+                raise InvalidGroupDefinition(
+                    'Group "%s" does not have any rules' % group_id)
+            for rule_id in rule_ids:
+                if rule_id not in self.rules_map:
+                    raise RuleDoesntExistError(
+                        'Rule id "%s" does not exist, cannot be in group' %
+                        rule_id)
+            self.rule_groups_map[group_id] = group_def.get(
+                'rule_ids', [])
+
+    def add_org_policy(self, org_def):
+        """Creates org policy and rule mapping.
+
+        Sample org structure:
+
+                org 1234
```

Comment:
Did you also add support for a root above the org?

Reply:
I did not, intentionally. I felt like the PR was getting a bit long so I was planning on adding that in the next iteration.